crates/wasmparser/src/parser.rs

```diff
@@ -1,1313 +1,1338 @@
+    ///             // afterwards we can parse and handle each function
+    ///             // individually.
+    ///             CodeSectionStart { .. } => { /* ... */ }
+    ///             CodeSectionEntry(body) => {
+    ///                 // here we can iterate over `body` to parse the function
+    ///                 // and its locals
+    ///             }
+    ///
+    ///             // When parsing nested modules we need to switch which
+    ///             // `Parser` we're using.
+    ///             ModuleCodeSectionStart { .. } => { /* ... */ }
+    ///             ModuleCodeSectionEntry { parser: subparser, .. } => {
+    ///                 stack.push(parser);
+    ///                 parser = subparser;
+    ///             }
+    ///
+    ///             CustomSection { name, .. } => { /* ... */ }
+    ///
+    ///             // most likely you'd return an error here
+    ///             UnknownSection { id, .. } => { /* ... */ }
+    ///
+    ///             // Once we've reached the end of a module we either resume
+    ///             // at the parent module or we break out of the loop because
+    ///             // we're done.
+    ///             End => {
+    ///                 if let Some(parent_parser) = stack.pop() {
+    ///                     parser = parent_parser;
+    ///                 } else {
+    ///                     break;
+    ///                 }
+    ///             }
+    ///         }
+    ///
+    ///         // once we're done processing the payload we can forget the
+    ///         // original.
+    ///         buf.drain(..consumed);
+    ///     }
+    ///
+    ///     Ok(())
+    /// }
+    ///
+    /// # parse(&b"\0asm\x01\0\0\0"[..]).unwrap();
     /// ```
-    pub fn new(data: &[u8]) -> Parser {
-        Parser {
-            data,
-            state: ParserState::Initial,
-            module_reader: None,
-            current_section: None,
-            section_reader: ParserSectionReader::None,
-            element_items: None,
-            current_function_body: None,
-            init_expr_continuation: None,
-            current_data_segment: None,
-            binary_reader: None,
-            operators_reader: None,
-            instance_args: None,
-            section_entries_left: 0,
-        }
-    }
-
-    pub fn eof(&self) -> bool {
-        match self.state {
-            ParserState::EndWasm => true,
-            ParserState::BeginWasm { .. } | ParserState::EndSection => {
-                self.module_reader.as_ref().expect("module reader").eof()
+    pub fn parse<'a>(&mut self, data: &'a [u8], eof: bool) -> Result<Chunk<'a>> {
+        let (data, eof) = if usize_to_u64(data.len()) > self.max_size {
+            (&data[..(self.max_size as usize)], true)
+        } else {
+            (data, eof)
+        };
+        // TODO: thread thread through `offset: u64` to `BinaryReader`, remove
+        // the cast here.
```

Comment:
:eyes:

Reply:
This is actually something I forgot to mention too. Lots of stuff in wasmparser is gearded towards `usize` in terms of offsets, but I think that we'll probably want to switch to `u64` now that they're not always relative to an in-memory stream but may instead be relative to a stream of bytes elsewhere.